File: src/main/java/io/junseok/mealmate/domain/comment/service/CommentService.java
package io.junseok.mealmate.domain.comment.service;

import io.junseok.mealmate.domain.board.entity.Board;
import io.junseok.mealmate.domain.board.service.BoardService;
import io.junseok.mealmate.domain.comment.dto.request.CommentWrite;
import io.junseok.mealmate.domain.comment.dto.response.CommentInfo;
import io.junseok.mealmate.domain.comment.entity.Comment;
import io.junseok.mealmate.domain.comment.repository.CommentRepository;
import io.junseok.mealmate.domain.member.entity.Member;
import io.junseok.mealmate.domain.member.service.MemberService;
import io.junseok.mealmate.exception.ErrorCode;
import io.junseok.mealmate.exception.MealMateException;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class CommentService {
    private final CommentRepository commentRepository;
    private final MemberService memberService;
    private final BoardService boardService;

    @Transactional
    public Long create(Long boardId, CommentWrite commentWrite, String email) {
        Member member = memberService.getMember(email);
        Board board = boardService.findBoard(boardId);
        Comment comment = Comment.builder()
            .commentContent(commentWrite.commentContent())
            .member(member)
            .board(board)
            .build();
        return commentRepository.save(comment).getCommentId();
    }

    @Transactional(readOnly = true)
    public List<CommentInfo> showComments(Long boardId) {
        Board board = boardService.findBoard(boardId);
        return commentRepository.findAllByBoard(board)
            .stream()
            .map(CommentInfo::fromEntity)
            .toList();
    }

    @Transactional
    public void updateComment(Long commentId, String email, String commentContent) {
        Comment comment = getComment(commentId);
        log.info(commentContent);
        if(!comment.getMember().getEmail().equals(email)){
            throw new MealMateException(ErrorCode.NOT_WRITE_MEMBER);
        }
        log.info(commentContent);
        comment.update(commentContent);
    }

    private Comment getComment(Long commentId) {
        return commentRepository.findById(commentId)
            .orElseThrow(() -> new MealMateException(ErrorCode.NOT_EXIST_COMMENT));
    }

    @Transactional
    public void remove(Long commentId, String email) {
        Comment comment = getComment(commentId);
        if(comment.getMember().getEmail().equals(email)){
            commentRepository.delete(comment);
        }
    }
}

File: src/main/java/io/junseok/mealmate/domain/comment/service/CommentService.kt
package io.junseok.mealmate.domain.comment.service

import io.junseok.mealmate.domain.board.service.BoardService
import io.junseok.mealmate.domain.comment.dto.request.CommentWrite
import io.junseok.mealmate.domain.comment.dto.response.CommentInfo
import io.junseok.mealmate.domain.comment.dto.response.toCreateCommentInfo
import io.junseok.mealmate.domain.comment.entity.Comment
import io.junseok.mealmate.domain.comment.repository.CommentRepository
import io.junseok.mealmate.domain.member.service.MemberService
import io.junseok.mealmate.exception.ErrorCode
import io.junseok.mealmate.exception.MealMateException
import lombok.extern.slf4j.Slf4j
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Slf4j
class CommentService(
    private val commentRepository: CommentRepository,
    private val memberService: MemberService,
    private val boardService: BoardService
) {
    @Transactional
    fun create(boardId: Long, commentWrite: CommentWrite, email: String): Long {
        val member = memberService.getMember(email)
        val board = boardService.findBoard(boardId)

        return Comment(
            commentContent = commentWrite.commentContent,
            member = member,
            board = board
        ).also {
            commentRepository.save(it)
        }.commentId!!
    }

    @Transactional(readOnly = true)
    fun showComments(boardId: Long): List<CommentInfo> {
        val board = boardService.findBoard(boardId)
        return commentRepository.findAllByBoard(board)
            .map { it.toCreateCommentInfo() }
    }

    @Transactional
    fun updateComment(commentId: Long, email: String, commentContent: String) {
        val comment = getComment(commentId)
        if (comment.member.email != email) {
            throw MealMateException(ErrorCode.NOT_WRITE_MEMBER)
        }
        comment.update(commentContent)
    }

    private fun getComment(commentId: Long): Comment =
        commentRepository.findByIdOrNull(commentId)
            ?: throw MealMateException(ErrorCode.NOT_EXIST_COMMENT)

    @Transactional
    fun remove(commentId: Long, email: String) {
        val comment = getComment(commentId)
        if (comment.member.email == email) {
            commentRepository.delete(comment)
        }
    }
}
